crates/polars-stream/src/nodes/parquet_source.rs

```diff
@@ -0,0 +1,1882 @@
+            .file_info
+            .reader_schema
+            .as_ref()
+            .unwrap()
+            .as_ref()
+            .unwrap_left()
+            .clone();
+
+        self.projected_arrow_fields =
+            if let Some(columns) = self.file_options.with_columns.as_deref() {
+                columns
+                    .iter()
+                    .map(|x| {
+                        // `index_of` on ArrowSchema is slow, so we use the polars native Schema,
+                        // but we need to remember to subtact the row index.
+                        let pos = self.file_info.schema.index_of(x.as_str()).unwrap()
+                            - (self.file_options.row_index.is_some() as usize);
+                        reader_schema.fields[pos].clone()
+                    })
+                    .collect()
+            } else {
+                Arc::from(reader_schema.fields.as_slice())
+            };
+
+        if self.verbose {
+            eprintln!(
+                "[ParquetSource]: {} columns to be projected from {} files",
+                self.projected_arrow_fields.len(),
+                self.paths.len(),
+            );
+        }
+    }
+}
+
+#[derive(Debug)]
+struct Config {
+    num_pipelines: usize,
+    /// Number of files to pre-fetch metadata for concurrently
+    metadata_prefetch_size: usize,
+    /// Number of files to decode metadata for in parallel in advance
+    metadata_decode_ahead_size: usize,
+    /// Number of row groups to pre-fetch concurrently, this can be across files
+    row_group_prefetch_size: usize,
+}
+
+/// Represents byte-data that can be transformed into a DataFrame after some computation.
+struct RowGroupData {
+    byte_source: FetchedBytes,
+    path_index: usize,
+    row_offset: usize,
+    slice: Option<(usize, usize)>,
+    file_max_row_group_height: usize,
+    row_group_metadata: RowGroupMetaData,
+    shared_file_state: Arc<tokio::sync::OnceCell<SharedFileState>>,
+}
+
+struct RowGroupDataFetcher {
+    metadata_rx: NotifyReceiver<(usize, usize, Arc<DynByteSource>, FileMetaData, usize)>,
+    use_statistics: bool,
+    verbose: bool,
+    reader_schema: Arc<ArrowSchema>,
+    projection: Option<Arc<[String]>>,
+    predicate: Option<Arc<dyn PhysicalIoExpr>>,
+    slice_range: Option<std::ops::Range<usize>>,
+    memory_prefetch_func: fn(&[u8]) -> (),
+    current_path_index: usize,
+    current_byte_source: Arc<DynByteSource>,
+    current_row_groups: std::vec::IntoIter<RowGroupMetaData>,
+    current_row_group_idx: usize,
+    current_max_row_group_height: usize,
+    current_row_offset: usize,
+    current_shared_file_state: Arc<tokio::sync::OnceCell<SharedFileState>>,
+}
+
+impl RowGroupDataFetcher {
+    fn into_stream(self) -> RowGroupDataStream {
+        let mut row_group_data_fetcher = Box::pin(self);
+        let current_future = Box::pin(
+            unsafe {
+                std::mem::transmute::<&mut RowGroupDataFetcher, &'static mut RowGroupDataFetcher>(
```

Comment:
Instead of solving this with lifetime transmutes which make me rather uncomfortable, can you change it to use `Arc` + `Mutex`es instead? You'll likely need to move the mutable state inside a `Mutex` and change `next()` to take `&self` instead of `&mut self`.

Reply:
I found a way to get rid of the unsafe transmute without introducing any locks 😁